### ISSUE TITLE
What do non-public/partner releases look like for .NET 8?

### ISSUE THREAD

**Issue by omajid** [dotnet/source-build#3717]
For .NET 6 and .NET 7, Microsoft makes a few things available to partners ahead of a public .NET release:

- A source code repository with the release tagged
- A source archive (.tar.gz file ) for .NET

With the .NET 8 model (VMR), the VMR public release looks like this:

- A source code repository (dotnet/dotnet) with the release tagged
- A source code archive (.tar.gz) auto-generated by github
- A release.json file

Does the non-public/partner release look the same? If not, what pieces are going to be available and how can we generate the rest?

For example, if there's a source code repository with the release tagged, I can generate a tar.gz using `git archive`. What can I do to make sure it matches the tarball that's available from GitHub when the release is public?

What about release.json? Will that be published? Will there be a way to generate something that works for everyone post-public-GA (that is, refer to the same commits on the public github repos)? 

How does the ability to verify releases (see https://github.com/dotnet/source-build/issues/3316) work for these partner releases?

**Comment by mmitche** [comment]
> For example, if there's a source code repository with the release tagged, I can generate a tar.gz using git archive. What can I do to make sure it matches the tarball that's available from GitHub when the release is public?

@omajid I struggled with this a bit around verifying the GitHub tarballs. I was unable to reliably generate a local git archive from our repos that exactly matches the GitHub tarball. The checksums of the tarballs always varied. From what I can tell, this probably has to do with the gzip and git versions that are in use on local machine vs. GitHub servers.

I was, however, able to unpack the tarballs and do directory comparisons on them (`diff -qr`) and get zero diff.

**Comment by mmitche** [comment]
@omajid We had a discussion about this in today's SB sync. Here is a proposal, and let me know what you think:

- For 8.0, we would add back the tarball to the pre-release day artifacts that partners have access to. This would just be a git archive generated from the partners repo.
- We would add a `release.json` alongside.
- On release day, you could verify the tarball you built against the public github release source tarball by doing a recursive directory compare of the two tarballs. We would still maintain the new signatures of the public tarball.

**Comment by omajid** [comment]
@mmitche This sounds sensible to me, given our constraints.

Would it be possible to produce a detached signature file on the partner side too? Then we (or anyone worried about authenticity of the source) could verify the parter-tarball using the partner-signature and the public-tarball using the public-signature.

**Comment by mmitche** [comment]
> @mmitche This sounds sensible to me, given our constraints. Thank you!
> 
> Would it be possible to produce a detached signature file on the partner side too? Then we (or anyone worried about authenticity of the source) could verify the parter-tarball using the partner-signature and the public-tarball using the public-signature.

Yes, we can do that.

**Comment by mmitche** [comment]
I updated my list above to include the partner signature

**Comment by MichaelSimons** [comment]
@oleksandr-didyk - please prioritize this so that it can be in place for the Feb servicing release.

**Comment by oleksandr-didyk** [comment]
PR for adding 8.0 VMR tarball to the release pipeline - https://dev.azure.com/dnceng/internal/_git/dotnet-release/pullrequest/36304

**Comment by oleksandr-didyk** [comment]
PR for adding the release manifest to the 8.0 VMR tarball - https://dev.azure.com/dnceng/internal/_git/dotnet-release/pullrequest/36485

**Comment by oleksandr-didyk** [comment]
> PR for adding the release manifest to the 8.0 VMR tarball - https://dev.azure.com/dnceng/internal/_git/dotnet-release/pullrequest/36485

Addition of release manifest currently blocked on another issue (https://github.com/dotnet/source-build/issues/3777), which in turn is waiting for a rollout on 01/24; after the rollout the addition of release manifest can be merged

**Comment by mmitche** [comment]
Same with the check-in of the automated signature generation.

**Comment by oleksandr-didyk** [comment]
> > PR for adding the release manifest to the 8.0 VMR tarball - [dev.azure.com/dnceng/internal/_git/dotnet-release/pullrequest/36485](https://dev.azure.com/dnceng/internal/_git/dotnet-release/pullrequest/36485)
> 
> Addition of release manifest currently blocked on another issue (#3777), which in turn is waiting for a rollout on 01/24; after the rollout the addition of release manifest can be merged

Blocking issue resolved, release manifest addition was merged. The 8.0 tarball and and accompanying release manifest should be available as of February release.

**Comment by MichaelSimons** [comment]
What work remains here?

**Comment by oleksandr-didyk** [comment]
All of the issues described here are resolved for Preview 2

**Comment by omajid** [comment]
From the March release, I have only one leftover minor concern about the difference between the github release vs the partner release:

The tarball name is different between the releases, and the tarball contents are nested slightly differently.

The GitHub release has a tarball named `dotnet-${tag_without_v_prefix}.tar.gz`, and all files in the tarball are nested under a directory named `dotnet-${tag_without_v_prefix}` (which is the same as the tarball without the `.tar.gz` suffix) 

```
$ tar tf dotnet-8.0.3.tar.gz  | head 
dotnet-8.0.3/
dotnet-8.0.3/.devcontainer/
dotnet-8.0.3/.devcontainer/README.md
dotnet-8.0.3/.devcontainer/devcontainer.json
dotnet-8.0.3/.devcontainer/init.sh
dotnet-8.0.3/.devcontainer/prebuilt-sdk/
dotnet-8.0.3/.devcontainer/prebuilt-sdk/devcontainer.json
dotnet-8.0.3/.devcontainer/prebuilt-sdk/init.sh
dotnet-8.0.3/.devcontainer/synchronize-vmr.sh
dotnet-8.0.3/.gitignore
```

The partner release is named `dotnet-sdk-source-${tag_without_v_prefix}.tar.gz`, and doesn't nest all files under a single directory:

```
$ tar tf dotnet-sdk-source-8.0.103.tar.gz | head
.devcontainer/
.devcontainer/README.md
.devcontainer/devcontainer.json
.devcontainer/init.sh
.devcontainer/prebuilt-sdk/
.devcontainer/prebuilt-sdk/devcontainer.json
.devcontainer/prebuilt-sdk/init.sh
.devcontainer/synchronize-vmr.sh
.gitignore
CODE_OF_CONDUCT.md
```

**Comment by oleksandr-didyk** [comment]
Thanks for pointing this out Omair, I'll take a look and correct the partner release tarball to follow the convention of the public release one

**Comment by oleksandr-didyk** [comment]
PR - https://dev.azure.com/dnceng/internal/_git/dotnet-release/pullrequest/38467

**Comment by oleksandr-didyk** [comment]
PR merged, should be fixed for `Preview 3`
